### ISSUE TITLE
Butterfly summation needed in cumsum

### ISSUE THREAD

**Issue by robparrishqc** [numpy/numpy#14909]
`numpy` (superb library - no slight intended here!) does not seem to support Kahan or butterfly summation in `cumsum`, which is critically important for accuracy in large arrays of numbers of similar magnitude. 

### Reproducing code example:

`import numpy as np; print(np.cumsum(np.ones((2**28,), dtype=np.float32))[-1] - 2**28)`

```python
import numpy as np
print(np.cumsum(np.ones((2**28,), dtype=np.float32))[-1] - 2**28)
# Should be zero
```

Note that `sum` obviously uses butterfly summation: 

```python
print(np.sum(np.ones((2**28,), dtype=np.float32)) - 2**28)
# Is zero on any machine I can find
```

**Comment by seberg** [comment]
Sum uses (partial) pairwise summation (in many cases). What is butterfly summation? There should be a few open issues about including stable summation methods in numpy, could you cross reference these here and comment there?

**Comment by robparrishqc** [comment]
I think butterfly summation == pairwise summation == parallel sum scan (just different names for the same thing). If there is any remaining doubt, the technique I am referring to is described extensively for CUDA here: https://developer.nvidia.com/gpugems/GPUGems3/gpugems3_ch39.html - this same approach should work quite well and be reasonably efficient for CPU, but some care will have to be taken to make sure the algorithm uses the cache structure efficiently (or you'll end up with something that scales like log2 memcpy ops). `np.sum` is clearly using something like this, but `np.cumsum` is clearly not (see example above). This is particularly problematic when summing large arrays of similar magnitude numbers in low precision (e.g., `float32`).

**Comment by seberg** [comment]
Yes, the issue is that `np.cumsum` uses `np.add.accumulate` which is a bit more limited in implementing special logic (at least right now).

If you can help out, I would prefer to move this discussion to an existing issue and close this one, although it may be that the existing stable summation issues are not specific about cumsum.

**Comment by robparrishqc** [comment]
Certainly - which open issue should I look at?

**Comment by charris** [comment]
I suppose one easy fix for float32 would be to accumulate the running sum in double precision.

**Comment by seberg** [comment]
True about float64 making the float32 issue less bad. xref gh-8786 I do not mind moving this there as well, but can also stay its own issue.

**Comment by nschloe** [comment]
Just an FTI: I once created [accupy](https://github.com/nschloe/accupy) for this purpose. It's slow though.
